File: samples/cext-trey-research-auth/src/services/ConsultantApiService.ts
import { Consultant, HoursEntry, Assignment } from '../model/baseModel';
import { ApiConsultant, ApiChargeTimeResponse } from '../model/apiModel';
import ProjectDbService from './ProjectDbService';
import AssignmentDbService from './AssignmentDbService';
import ConsultantDbService from './ConsultantDbService';
import { HttpError } from './Utilities';
import ProjectApiService from './ProjectApiService';

const AVAILABLE_HOURS_PER_MONTH = 160;

class ConsultantApiService {

    async getApiConsultantById(consultantId: string): Promise<ApiConsultant> {

        let result = null;
        let consultant = await ConsultantDbService.getConsultantById(consultantId);
        if (consultant) {
            let assignments = await AssignmentDbService.getAssignments();
            result = await this.getApiConsultantForBaseConsultant(consultant, assignments);
        }
        return result;
    }

    async getApiConsultants(
        consultantName: string, projectName: string, skill: string,
        certification: string, role: string, hoursAvailable: string): Promise<ApiConsultant[]> {

        let consultants = await ConsultantDbService.getConsultants();
        let assignments = await AssignmentDbService.getAssignments();

        // Filter on base properties
        if (consultantName) {
            consultants = consultants.filter(
                (c) => c.name.toLowerCase().includes(consultantName.toLocaleLowerCase()));
        }
        if (skill) {
            consultants = consultants.filter(
                (c) => c.skills.find((s) => s.toLowerCase().includes(skill.toLocaleLowerCase())));
        }
        if (certification) {
            consultants = consultants.filter(
                (c) => c.certifications.find((s) => s.toLowerCase().includes(certification.toLocaleLowerCase())));
        }
        if (role) {
            consultants = consultants.filter(
                (c) => c.roles.find((s) => s.toLowerCase().includes(role)));
        }

        // Augment the base properties with assignment information
        let result = await Promise.all(consultants.map((c) => this.getApiConsultantForBaseConsultant(c, assignments)));

        // Filter on project name
        if (result && projectName) {
            result = result.filter(
                (c) => {
                    let project = c.projects.find((p) => {
                        let x = p.projectName.toLowerCase() + p.clientName.toLowerCase();
                        return x.includes(projectName);
                    });
                    return project;
                });
        };
        // Filter on available hours
        if (result && hoursAvailable) {
            result = result.filter(
                (c) => {
                    let availableHours = AVAILABLE_HOURS_PER_MONTH * 2 - c.forecastThisMonth - c.forecastNextMonth;
                    return availableHours >= parseInt(hoursAvailable);
                });
        };

        return result;
    }

    public async createApiConsultant(consultant: Consultant): Promise<ApiConsultant> {
        await ConsultantDbService.createConsultant(consultant);
        const assignments = await AssignmentDbService.getAssignments();

        const newApiConsultant = 
            this.getApiConsultantForBaseConsultant(consultant, assignments);
        return newApiConsultant;
    }

    // Augment a base consultant to get an ApiConsultant
    async getApiConsultantForBaseConsultant(consultant: Consultant, assignments: Assignment[]): Promise<ApiConsultant> {

        const result = {
            id: consultant.id,
            name: consultant.name,
            email: consultant.email,
            phone: consultant.phone,
            consultantPhotoUrl: consultant.consultantPhotoUrl,
            location: consultant.location,
            skills: consultant.skills,
            certifications: consultant.certifications,
            roles: consultant.roles,
            projects: [],
            forecastThisMonth: 0,
            forecastNextMonth: 0,
            deliveredLastMonth: 0,
            deliveredThisMonth: 0
        }
        assignments = assignments.filter((a) => a.consultantId === consultant.id);

        result.forecastThisMonth = 0;
        result.forecastNextMonth = 0;
        result.deliveredLastMonth = 0;
        result.deliveredThisMonth = 0;

        for (let assignment of assignments) {
            const project = await ProjectDbService.getProjectById(assignment.projectId);
            const { lastMonthHours: forecastLastMonth,
                thisMonthHours: forecastThisMonth,
                nextMonthHours: forecastNextMonth } = this.findHours(assignment.forecast);
            const { lastMonthHours: deliveredLastMonth,
                thisMonthHours: deliveredThisMonth,
                nextMonthHours: deliveredNextMonth } = this.findHours(assignment.delivered);

            result.projects.push({
                projectName: project.name,
                projectDescription: project.description,
                projectLocation: project.location,
                mapUrl: project.mapUrl,
                clientName: project.clientName,
                clientContact: project.clientContact,
                clientEmail: project.clientEmail,
                role: assignment.role,
                forecastThisMonth: forecastThisMonth,
                forecastNextMonth: forecastNextMonth,
                deliveredLastMonth: deliveredLastMonth,
                deliveredThisMonth: deliveredThisMonth
            });
            result.forecastThisMonth += forecastThisMonth;
            result.forecastNextMonth += forecastNextMonth;
            result.deliveredLastMonth += deliveredLastMonth;
            result.deliveredThisMonth += deliveredThisMonth;

        }
        return result;
    }

    // Extract this and next month's hours from an array of HoursEntry
    private findHours(hours: HoursEntry[]): { lastMonthHours: number, thisMonthHours: number, nextMonthHours: number } {
        const now = new Date();
        const thisMonth = now.getMonth();
        const thisYear = now.getFullYear();

        const lastMonth = thisMonth === 0 ? 11 : thisMonth - 1;
        const lastYear = thisMonth === 0 ? thisYear - 1 : thisYear;

        const nextMonth = thisMonth === 11 ? 0 : thisMonth + 1;
        const nextYear = thisMonth === 11 ? thisYear + 1 : thisYear;

        const result = {
            lastMonthHours: hours.find((h) => h.month === lastMonth + 1 && h.year === lastYear)?.hours || 0,
            thisMonthHours: hours.find((h) => h.month === thisMonth + 1 && h.year === thisYear)?.hours || 0,
            nextMonthHours: hours.find((h) => h.month === nextMonth + 1 && h.year === nextYear)?.hours || 0
        };
        return result;
    }

    async chargeTimeToProject(projectName: string, consultantId: string, hours: number): Promise<ApiChargeTimeResponse> {
        let projects = await ProjectApiService.getApiProjects(projectName, "");
        if (projects.length === 0) {
            throw new HttpError(404, `Project not found: ${projectName}`);
        } else if (projects.length > 1) {
            throw new HttpError(406, `Multiple projects found with the name: ${projectName}`);
        } else {
            const project = projects[0];
            // Always charge to the current month
            const month = new Date().getMonth() + 1;
            const year = new Date().getFullYear();
            const remainingForecast = await AssignmentDbService.chargeHoursToProject(project.id, consultantId, month, year, hours);
            let message = "";
            if (remainingForecast < 0) {
                message = `Charged ${hours} hours to ${project.clientName} on project "${project.name}". You are ${-remainingForecast} hours over your forecast this month.`;
            } else {
                message = `Charged ${hours} hours to ${project.clientName} on project "${project.name}". You have ${remainingForecast} hours remaining this month.`;
            }
            return {
                clientName: project.clientName,
                projectName: project.name,
                remainingForecast,
                message
            };
        }
    }
}

export default new ConsultantApiService();

File: samples/cext-trey-research-python/services/consultant_api_service.py
from typing import List, Optional
from datetime import datetime
from model.base_model import Consultant, HoursEntry, Assignment
from model.api_model import ApiConsultant, ApiChargeTimeResponse
from services.project_db_service import project_db_service
from services.assignment_db_service import assignment_db_service
from services.consultant_db_service import consultant_db_service
from .utilities import HttpError
import asyncio

AVAILABLE_HOURS_PER_MONTH = 160

class ConsultantApiService:

    async def get_api_consultant_by_id(self, consultant_id: str) -> Optional[ApiConsultant]:
        result = None
        consultant = await consultant_db_service.get_consultant_by_id(consultant_id)
        if consultant:
            assignments = await assignment_db_service.get_assignments()
            result = await self.get_api_consultant_for_base_consultant(consultant, assignments)
        return result

    async def get_api_consultants(self, consultant_name: str, project_name: str, skill: str,
                                  certification: str, role: str, hours_available: str) -> List[ApiConsultant]:

        consultants = await consultant_db_service.get_consultants()
        assignments = await assignment_db_service.get_assignments()

        # Filter on base properties
        if consultant_name:
            consultants = [c for c in consultants if consultant_name.lower() in c.get("name").lower()]
        if skill:
            consultants = [c for c in consultants if any(skill.lower() in s.lower() for s in c.get("skills"))]
        if certification:
            consultants = [c for c in consultants if any(certification.lower() in cert.lower() for cert in c.get("certifications"))]
        if role:
            consultants = [c for c in consultants if role.lower() in [r.lower() for r in c.get("roles")]]

        # Augment the base properties with assignment information
        result = await asyncio.gather(*[self.get_api_consultant_for_base_consultant(c, assignments) for c in consultants])

        # Filter on project name
        if project_name:
            result = [c for c in result if any(
                project_name.lower() in (p.get("projectName").lower() + p.get("clientName").lower()) for p in c.projects
            )]

        # Filter on available hours
        if hours_available:
            hours_available = int(hours_available)
            result = [c for c in result if AVAILABLE_HOURS_PER_MONTH * 2 - c.forecast_this_month - c.forecast_next_month >= hours_available]

        return result

    async def create_api_consultant(self, consultant: Consultant) -> ApiConsultant:
        await consultant_db_service.create_consultant(consultant)
        assignments = await assignment_db_service.get_assignments()

        new_api_consultant = await self.get_api_consultant_for_base_consultant(consultant, assignments)
        return new_api_consultant

    async def get_api_consultant_for_base_consultant(self, consultant: Consultant, assignments: List[Assignment]) -> ApiConsultant:
        result = ApiConsultant(
            id=consultant.get('id'),
            name=consultant.get('name'),
            email=consultant.get('email'),
            phone=consultant.get('phone'),
            consultantPhotoUrl=consultant.get('consultantPhotoUrl'),
            location=consultant.get('location'),
            skills=consultant.get('skills'),
            certifications=consultant.get('certifications'),
            roles=consultant.get('roles'),
            projects=[],  
            forecastThisMonth=0,
            forecastNextMonth= 0,
            deliveredLastMonth= 0,
            deliveredThisMonth= 0
        )

        assignments = [a for a in assignments if a.consultantId == consultant.get('id')]

        for assignment in assignments:
            project = await project_db_service.get_project_by_id(assignment.projectId)
            forecast_hours = self.find_hours(assignment.forecast)
            delivered_hours = self.find_hours(assignment.delivered)
            result.projects.append({
                "projectName": project.name,
                "projectDescription": project.description,
                "projectLocation": project.location,
                "mapUrl": project.mapUrl,
                "clientName": project.clientName,
                "clientContact": project.clientContact,
                "clientEmail": project.clientEmail,
                "role": assignment.role,
                "forecastThisMonth": forecast_hours["thisMonthHours"],
                "forecastNextMonth": forecast_hours["nextMonthHours"],
                "deliveredLastMonth": delivered_hours["lastMonthHours"],
                "deliveredThisMonth": delivered_hours["thisMonthHours"]
            })

            result.forecastThisMonth += forecast_hours["thisMonthHours"]
            result.forecastNextMonth += forecast_hours["nextMonthHours"]
            result.deliveredLastMonth += delivered_hours["lastMonthHours"]
            result.deliveredThisMonth += delivered_hours["thisMonthHours"]

        return result

    def find_hours(self, hours: List[HoursEntry]) -> dict:
        now = datetime.now()
        this_month = now.month
        this_year = now.year

        last_month = 12 if this_month == 1 else this_month - 1
        last_year = this_year - 1 if this_month == 1 else this_year

        next_month = 1 if this_month == 12 else this_month + 1
        next_year = this_year + 1 if this_month == 12 else this_year
        return {
            "lastMonthHours": next((h.get('hours') for h in hours if h.get('month') == last_month and h.get('year') == last_year), 0),
            "thisMonthHours": next((h.get('hours') for h in hours if h.get('month') == this_month and h.get('year') == this_year), 0),
            "nextMonthHours": next((h.get('hours') for h in hours if h.get('month') == next_month and h.get('year') == next_year), 0)
        }

    async def charge_time_to_project(self, project_name: str, consultant_id: str, hours: int) -> ApiChargeTimeResponse:
        from services.project_api_service import project_api_service
        projects = await project_api_service.get_api_projects(project_name, "")
        if not projects:
            raise HttpError(404, f"Project not found: {project_name}")
        if len(projects) > 1:
            raise HttpError(406, f"Multiple projects found with the name: {project_name}")
        
        project = projects[0]

        # Always charge to the current month
        month = datetime.now().month
        year = datetime.now().year

        remaining_forecast = await assignment_db_service.charge_hours_to_project(project.id, consultant_id, month, year, hours)

        if remaining_forecast < 0:
            message = f"Charged {hours} hours to {project.clientName} on project \"{project.name}\". You are {-remaining_forecast} hours over your forecast this month."
        else:
            message = f"Charged {hours} hours to {project.clientName} on project \"{project.name}\". You have {remaining_forecast} hours remaining this month."

        return ApiChargeTimeResponse(
            clientName=project.clientName,
            projectName=project.name,
            remainingForecast=remaining_forecast,
            message=message
        )

# Export instance of ConsultantApiService
consultant_api_service = ConsultantApiService()